### PR TITLE
Rover: move more vehicle code into mode class

### PR DESCRIPTION
**Pull request by rmackay9**
This PR moves most of the navigation code out of the vehicle code and into the new mode classes.

- set_desired_location, set_desired_heading_and_speed are moved into the parent Mode class so that they can be used by other modes (guided, rtl, auto)
- get_distance_to_destination, speed_error provide info back to the vehicle code for reporting to the GCS and logging
- auto gets a HeadingAndSpeed submode which is used to replace handling of CONDITION_YAW (which is a do-command that was improperly being used as a navigation command) with SET_YAW_SPEED (a nav command only used by ArduPilot rover)
- guided mode gets improved handling of set-attitude-target, set-position-target messages including support for yaw and yaw-rate fields.
- removal of a few variables, functions and files from Rover now that their functionality is moved to the mode classes.

These changes are mostly non-functional but there are a few things to be aware of:

- RTL no longer switches to HOLD when it arrives at home.  It is no longer necessary because RTL will stop on it's own now.  It is fairly easy to add back in the change to HOLD mode but personally I prefer not to change modes without user input.
- SET_YAW_SPEED handling is slightly inconsistent with the mavlink command-long definition (throttle is -1 ~ +1, angle is in degrees, new relative-angle field added) but we are the only users of this message so we can just update the mavlink spec.

This has been tested in the simulator and on my skid-steering and regular style vehicles.

### CODE REVIEW

File: APMrover2/APMrover2.cpp [hunk 1]
```diff
@@ -180,6 +177,8 @@ void Rover::ahrs_update()
     Vector3f velocity;
     if (ahrs.get_velocity_NED(velocity)) {
         ground_speed = norm(velocity.x, velocity.y);
+    } else if (gps.status() >= AP_GPS::GPS_OK_FIX_3D) {
```
Comment (peterbarker) [inline APMrover2/APMrover2.cpp]: FIX_2D might be OK here.

File: APMrover2/APMrover2.cpp [hunk 2]
```diff
@@ -392,10 +391,6 @@ void Rover::update_GPS_10Hz(void)
         set_system_time_from_GPS();
 
         if (gps.status() >= AP_GPS::GPS_OK_FIX_3D) {
```
Comment (peterbarker) [inline APMrover2/APMrover2.cpp]: This check is actually redundant; AP_Camera's update function checks this first-thing (for right or wrong).

File: APMrover2/GCS_Mavlink.cpp [hunk 3]
```diff
@@ -1255,27 +1256,45 @@ void GCS_MAVLINK_Rover::handleMessage(mavlink_message_t* msg)
...
+                    target_yaw_cd = wrap_180_cd(target_yaw_cd + rover.ahrs.yaw_sensor);
                 }
-            } else if (pos_ignore && !vel_ignore && acc_ignore) {
-                rover.set_guided_velocity(target_steer_speed, target_speed);
-            } else if (!pos_ignore && vel_ignore && acc_ignore) {
-                rover.set_guided_WP(target_loc);
-            } else {
-                // result = MAV_RESULT_FAILED;  // TODO : support that
+            }
+            if (!yaw_rate_ignore) {
```
Comment (peterbarker) [inline APMrover2/GCS_Mavlink.cpp]: There are ~30 lines of duplicated code here.  Would be nice to get a subroutine.

File: APMrover2/mode.cpp [hunk 4]
```diff
@@ -29,87 +25,63 @@ bool Mode::enter()
...
-    float reduction = 1.0f - steer_rate * speed_turn_reduction;
-
-    if (is_autopilot_mode() && rover.mode_guided.guided_mode != ModeGuided::Guided_Velocity && wp_distance <= g.speed_turn_dist) {
-        // in auto-modes we reduce speed when approaching waypoints
-        const float reduction2 = 1.0f - speed_turn_reduction;
-        if (reduction2 < reduction) {
-            reduction = reduction2;
-        }
-    }
+// set desired heading and speed
```
Comment (peterbarker) [inline APMrover2/mode.cpp]: Maybe add a comment saying it's only relevant in auto modes?

File: APMrover2/APMrover2.cpp (hunk 2 above)
Comment (khancyr) [inline APMrover2/APMrover2.cpp]: yep I already correct this one ! just need to push the patch

File: APMrover2/APMrover2.cpp (hunk 1 above)
Comment (rmackay9) [inline APMrover2/APMrover2.cpp]: Perhaps, but that is the check used in master at the moment..

File: APMrover2/APMrover2.cpp (hunk 2 above)
Comment (rmackay9) [inline APMrover2/APMrover2.cpp]: sounds good to remove. it's an existing issue of course, not related to this patch (except that this patch happens to touch lines nearby).

File: APMrover2/mode.cpp (hunk 4 above)
Comment (rmackay9) [inline APMrover2/mode.cpp]: sure, added.

**Pull request merged**